tls/_common/prf.py

```diff
@@ -0,0 +1,75 @@
+# This file is dual licensed under the terms of the Apache License, Version
+# 2.0, and the BSD License. See the LICENSE file in the root of this repository
+# for complete details.
+
+from __future__ import absolute_import, division, print_function
+
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.primitives.hmac import HMAC
+
+from tls.exceptions import UnsupportedHashAlgorithm
+
+
+def _p_hash(hash_algorithm, secret, seed, output_length):
+    """
+    A seed expansion function that uses a single hash function to expand a
+    secret and seed into the number of bytes specified by output_length.
+    """
+    result = b""
+    i = 1
+    while len(result) < output_length:
+        h = HMAC(secret, hash_algorithm, default_backend())
+        h.update(_a(secret, hash_algorithm, i, seed))
+        h.update(seed)
+        result += h.finalize()
+        i += 1
+    return result[:output_length]
+
+
+def _a(secret, hash_algorithm, n, seed):
+    """
+    a() is defined as:
+        a(0) = seed
+        a(i) = HMAC_hash(secret, A(i-1))
+    """
+    if n == 0:
+        return seed
+    else:
+        h = HMAC(secret, hash_algorithm, default_backend())
+        h.update(_a(secret, hash_algorithm, n - 1, seed))
+        return h.finalize()
+
+
+def prf(secret, label, seed, hash_algorithm, output_length):
+    """
+    A construction to expand secrets into blocks of data for the purposes of
+    key generation or validation.
+
+    This pseudorandom function (PRF) takes as input a secret, a seed, and an
+    identifying label and produces an output of length specified in
+    output_length.
+
+    :param secret: Secret key as ``bytes``.  The key should be randomly
+    generated bytes and is recommended to be equal in length to the digest_size
+    of the hash function chosen. You must keep the key secret.
+
+    :param label: An ASCII string.
+
+    :param seed: The seed as ``bytes``.
+
+    :param hash_algorithm: An
+    ``cryptography.hazmat.primitives.hashes.HashAlgorithm`` provider.
+
+    :param output_length: The number of bytes to expand the seed into.
+    """
+    # The TLS PRF must be used with SHA-256 or a stronger standard hash
+    # function.
+    supported_hash_algorithms = frozenset([hashes.SHA256.name])
```

Comment:
I could use help figuring what should go in here. After talking to @reaperhulk and @glyph about this, we also figured that it is OK if more validation here is added and/or the set is expanded later on.